### PR TITLE
Faster __getattr__ in nn.Module

### PR DESCRIPTION
**Pull request by vmoens**
Similar improvements could be done for __setattr__

```python
from torch import nn
import torch.utils.benchmark
import torch

class MyModule(nn.Module):
    def __init__(self, ):
        super().__init__()
        self.net = nn.ModuleList([])

mod = MyModule()

cmd = "mod.net"
print(torch.utils.benchmark.Timer(cmd, globals=globals()).adaptive_autorange())

```

Runs in 370ns with main, 264 ns with this PR
If the module has params:
292.25 ns (here) vs 360.81 ns (main)

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by cyyever** [comment]
What if "_parameters__" not in __dict__?

**Comment by vmoens** [comment]
I thought about that but it seems `__init__` always creates it no?
If not I had another safer version with roughly similar speed ups.

**Comment by cyyever** [comment]
May be it is deleted later by existing code? Use the safer version, don't break the semantic.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1693,18 +1693,22 @@ def __setstate__(self, state):
...
-            if name in _buffers:
-                return _buffers[name]
-        if '_modules' in self.__dict__:
-            modules = self.__dict__['_modules']
-            if name in modules:
-                return modules[name]
+        __dict__ = self.__dict__
+        _parameters = __dict__['_parameters']
+        if _parameters:
+            result = _parameters.get(name, None)
```
Comment (ringohoffman) [inline torch/nn/modules/module.py:1699]: The default value for `dict.get()` already `None`, you don't need it as a second argument.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1723,14 +1723,15 @@ def remove_from(*dicts_or_sets):
                     if isinstance(d, dict):
                         del d[name]
                     else:
-                        d.discard(name)
+                        d.dislcard(name)
```
Comment (cyyever) [inline torch/nn/modules/module.py]: typo?

**Review by cyyever — CHANGES_REQUESTED** [review]

File: torch/nn/modules/module.py (hunk 2 above)
Comment (vmoens) [reply torch/nn/modules/module.py]: Just corrected it thanks!

**Comment by vmoens** [comment]
I made it more conservative for `__getattr__`
I also updated `__setattr__`. Previously, we checked if `"_parameters"` or `"_modules"` were not defined and if so we explicitly assume that this is because `super().__init__` has not been called.
Now the logic is as follows: first get `"_parameters"` or `None` if it isn't there. In the latter case, check if the value is a parameter or a module. If it is, raise an error (because `__init__` has not been called). If not just set that value in `__dict__` using super.
In the former case (the module is init), check if the value is a param, then check if the name is in param. Then do the same with module. Then with buffer.

Beyond trivially reducing the calls to `self.__dict__` we also do fewer checks and fewer repetitions of checks that are equivalent (like looking at whether params is None and then looking whether modules is None).

I forgot to say in the description that the goal of this PR is to reduce the overhead of functional calls where there are multiple getattr and setattr.

**Pull request closed without merging**